clients/venice-thin-client/src/main/java/com/linkedin/venice/client/schema/RouterBackedSchemaReader.java

```diff
@@ -253,29 +351,38 @@ private void refreshAllValueSchema() throws VeniceClientException {
             schemaId);
         valueSchemaMap.put(schemaId, writerSchema);
         valueSchemaMapR.put(writerSchema, schemaId);
-        if (latestSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID || latestSchemaId < schemaId) {
-          latestSchemaId = schemaId;
+        if (maxSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID || maxSchemaId < schemaId) {
+          maxSchemaId = schemaId;
         }
-        if (preferredSchemaFilter.test(writerSchema)
-            && (latestPreferredSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID || latestPreferredSchemaId < schemaId)) {
-          latestPreferredSchemaId = schemaId;
+        if (preferredSchemaFilter.test(writerSchema)) {
+          if (schemaId == supersetSchemaId) {
+            supersetSchemaIsPreferredSchema = true;
+          }
+
+          if (maxPreferredSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID || maxPreferredSchemaId < schemaId) {
+            maxPreferredSchemaId = schemaId;
+          }
         }
       }
```

Comment:
This piece of logic seems pretty confusing because of these flags check in the for loop and after the loop..
Can we divide the logic into two parts?
1. Superset schema has the higher priority, so let us check superset schema first and if it can pass these checks, then just update the latest value schema.
2. If superset schema doesn't exist or doesn't pass the check, then iterate the loop and find out the preferred schema.

Do you think this will help simplify the logic here? At least it will move superset schema handling out of this loop/if-else branches?

Reply:
I've tried to make the code a bit more readable. Let me know if you'd like me to try make more changes